compiler/rustc_parse/src/parser/pat.rs

```diff
@@ -446,12 +447,195 @@ impl<'a> Parser<'a> {
+                                LocalKind::Decl | LocalKind::Init(_) => {
+                                    err.subdiagnostic(
+                                        UnexpectedExpressionInPatternSugg::RemoveLet {
+                                            let_span: local
+                                                .span
+                                                .shrink_to_lo()
+                                                .until(local.pat.span),
+                                            ty_span: local
+                                                .colon_sp
+                                                .map(|sp| sp.to(local.ty.as_ref().unwrap().span)),
+                                            has_initializer: !matches!(local.kind, LocalKind::Decl),
+                                        },
+                                    );
+
+                                    // don't suggest `let const { pat } = expr;`
+                                    return;
+                                }
+
+                                LocalKind::InitElse(_, _) => {}
+                            }
+                        }
+
+                        // help: use an arm guard `if val == expr`
+                        if let Some(arm) = &self.arm
+                            && !is_range_bound
+                        {
+                            let (ident, ident_span) = match self.field {
+                                Some(field) => {
+                                    (field.ident.to_string(), field.ident.span.to(expr_span))
+                                }
+                                None => ("val".to_owned(), expr_span),
+                            };
+
+                            match &arm.guard {
+                                None => {
+                                    err.subdiagnostic(
+                                        UnexpectedExpressionInPatternSugg::CreateGuard {
+                                            ident_span,
+                                            pat_hi: arm.pat.span.shrink_to_hi(),
+                                            ident,
+                                            expr: expr.clone(),
+                                        },
+                                    );
+                                }
+                                Some(guard) => {
+                                    err.subdiagnostic(
+                                        UnexpectedExpressionInPatternSugg::UpdateGuard {
+                                            ident_span,
+                                            guard_lo: guard.span.shrink_to_lo(),
+                                            guard_hi: guard.span.shrink_to_hi(),
+                                            ident,
+                                            expr: expr.clone(),
+                                        },
+                                    );
```

Comment:
In a future PR, we should probably make `CreateGuard` and `UpdateGuard` not use `expr` to achieve a 'cleaner'/more legible suggestion by not relying on `span_to_snippet`.

Rn, the suggestion copy/pastes the original expr into the suggestion:

```
3 |         val if val == 1 + 2 => {}
  |         ~~~ +++++++++++++++
```

if we just suggested only `val if val == ` instead, it would be

```
3 |         val if val == 1 + 2 => {}
  |         +++++++++++++
```

(and if we need to wrap the expr due to precedence, we can just use a multipart suggestion: `val if val == `, `)`)

It's always a win to rely less on `span_to_snippet` which is fallible (CC #128442, #128445).